### PR TITLE
txnkv: add new API for lock->put optimization

### PR DESCRIPTION
**Pull request by zyguan**

### CODE REVIEW

File: txnkv/transaction/2pc.go [hunk 1]
```diff
@@ -508,15 +523,25 @@ func (c *twoPhaseCommitter) initKeysAndMutations(ctx context.Context) error {
 		_ = err
 		key := it.Key()
 		flags := it.Flags()
-		var value []byte
-		var op kvrpcpb.Op
+		var (
+			value       []byte
+			cachedValue []byte
+			op          kvrpcpb.Op
+		)
```
Comment (cfzjywxk) [inline txnkv/transaction/2pc.go:536]: I think it's safer to add a flag to indicate if the value is from the `lock to put`,
```
valueFromLockToPut := false
```

File: txnkv/transaction/2pc.go [hunk 2]
```diff
@@ -581,8 +606,8 @@ func (c *twoPhaseCommitter) initKeysAndMutations(ctx context.Context) error {
 		if c.txn.schemaAmender != nil || c.txn.assertionLevel == kvrpcpb.AssertionLevel_Off {
 			mustExist, mustNotExist, hasAssertUnknown = false, false, false
 		}
-		c.mutations.Push(op, isPessimistic, mustExist, mustNotExist, it.Handle())
-		size += len(key) + len(value)
+		c.mutations.Push(op, isPessimistic, mustExist, mustNotExist, it.Handle(), cachedValue)
```
Comment (cfzjywxk) [inline txnkv/transaction/2pc.go:615]: Then use the `valueFromLockToPut` mentioned above as the input parameter.

File: txnkv/transaction/2pc.go [hunk 3]
```diff
@@ -252,6 +261,12 @@ func (m *memBufferMutations) Push(op kvrpcpb.Op, isPessimisticLock, assertExist,
 	}
 	handle.UserData = aux
 	m.handles = append(m.handles, handle)
+	if value != nil {
```
Comment (cfzjywxk) [inline txnkv/transaction/2pc.go]: And then here
```
if valueFromLockToPut {
   ...
}
```

Maybe it's safer than checking the `nil` reference.

File: integration_tests/2pc_test.go [hunk 4]
```diff
@@ -1742,3 +1743,90 @@ func (s *testCommitterSuite) TestFlagsInMemBufferMutations() {
...
+		},
+		{
+			"LockAndDelete",
+			[]func(txn transaction.TxnProbe){
+				func(txn transaction.TxnProbe) { txn.SetLockedKeyValue(k1, v1) },
+				func(txn transaction.TxnProbe) { mustLockKey(txn, k1) },
+				func(txn transaction.TxnProbe) { s.Require().NoError(txn.Delete(k1)) },
+			},
+			mutOpVals(kvrpcpb.Op_Del, []byte{}),
+		},
```
Comment (MyonKeminta) [inline integration_tests/2pc_test.go:1813]: Is it necessary to test in different order, like calling `txn.SetLockedKeyValue` after calling `LockKeys` or any mutations?

**Review by MyonKeminta — APPROVED** [review]
LGTM

File: integration_tests/2pc_test.go (hunk 4 above)
Comment (MyonKeminta) [reply integration_tests/2pc_test.go:1813]: Ah, I submitted the  review to wrong place

**Pull request closed without merging**